ouroboros-consensus/test-consensus/Test/Dynamic/General.hs

```diff
@@ -337,3 +360,91 @@ prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} schedule
+    -- they were forged. However, some delays are expected, due to nodes
+    -- joining late and also due to the practicality of the ChainSync and
+    -- BlockFetch policies, which try to avoid /unnecessary/ header/block
+    -- fetches. See the relevant comments below.
+    --
+    -- NOTE: This current property does not check for interminable message
+    -- delay: i.e. for blocks that were never added to some ChainDBs. It only
+    -- checks the slot difference once a message does arrive. This seems
+    -- acceptable: if there are no Common Prefix or Chain Growth violations,
+    -- then each message must have either arrived or ultimately been
+    -- irrelevant.
+    --
+    prop_no_unexpected_message_delays :: HasCallStack => Property
+    prop_no_unexpected_message_delays =
+        conjoin $
+        [ case p of
+              GenesisPoint            -> error "impossible"
+              BlockPoint sendSlot hsh ->
+                  prop1 nid recvSlot sendSlot hsh bno
+        | (nid, m)          <- Map.toList adds
+        , (recvSlot, pbnos) <- Map.toList m
+        , (p, bno)          <- Set.toList pbnos
+        ]
+      where
+        -- INVARIANT: these AddBlock events are *not* for EBBs
+        adds = nodeOutputAdds <$> testOutputNodes
+
+        prop1 nid recvSlot sendSlot hsh bno =
+            counterexample msg $
+            delayOK || noDelay
+          where
+            msg =
+                "Unexpected message delay " <>
+                "(" <> "recipient: " <> condense nid <>
+                "," <> "expected receive slot: "
+                    <> condense firstPossibleReception <>
+                "," <> "actual receive slot: " <> condense recvSlot <>
+                "," <> "blockHash: " <> show hsh <>
+                "," <> "blockNo: " <> condense (unBlockNo bno) <>
+                ")"
+
+            -- a node cannot receive a block until both exist
+            firstPossibleReception =
+                nodeIdJoinSlot nodeJoinPlan nid `max` sendSlot
+
+            noDelay = recvSlot == firstPossibleReception
+
+            delayOK = delayOK1 || delayOK2
+            -- When nodes lead in the same slot in which they join the network,
+            -- they immediately forge a single block on top of Genesis; this
+            -- block then prevents them from fetching the network's current
+            -- chain if it also consists of just one block.
+            --
+            -- NOTE This predicate is more general than that specific scenario,
+            -- but we don't anticipate it wholly masking any interesting cases.
+            delayOK1 = 1 == bno
+
+            -- When a slot has multiple leaders, each node chooses one of the
+            -- mutually-exclusive forged blocks and won't fetch any of the
+            -- others until it's later compelled to switch to a chain
+            -- containing one of them
+            --
+            -- TODO This predicate is more general than that specific scenario,
```

Comment:
I'm tempted to depriortize this TODO, since it'll be superceded by the eventual PR that introduces network latency injections.

Reply:
I amended the comment to mention the latency-inducing issues that we expect will handle it.